File: src/ast/expr.rs
// src/ast/expr.rs

use crate::ast::method::Method;

#[derive(Debug, Clone, PartialEq)]
pub enum ExprKind<'a> {
    LiteralInt(i32),
    LiteralLong(i64),
    LiteralFloat(f32),
    LiteralDouble(f64),
    LiteralString(&'a str),
    LiteralChar(char),
    LiteralBool(bool),
    LiteralNopaw,

    Var(&'a str),

    UnaryOp {
        op: &'a str,
        expr: Box<Expr<'a>>,
    },

    BinaryOp {
        op: BinaryOp,
        left: Box<Expr<'a>>,
        right: Box<Expr<'a>>,
    },

    Call {
        name: &'a str,
        args: Vec<Expr<'a>>,
    },

    MethodCall {
        receiver: Box<Expr<'a>>,
        method: Method,
        args: Vec<Expr<'a>>,
    },

    Cast {
        expr: Box<Expr<'a>>,
        ty: &'a str,
    },

    ArrayLiteral(Vec<Expr<'a>>),
    Index {
        array: Box<Expr<'a>>,
        index: Box<Expr<'a>>,
    },
    FieldAccess {
        expr: Box<Expr<'a>>,
        field: &'a str,
    },
    RecordInit {
        name: &'a str,
        fields: Vec<(&'a str, Expr<'a>)>,
    },
    Await {
        expr: Box<Expr<'a>>,
    },
}

/// 带位置的表达式
#[derive(Debug, Clone, PartialEq)]
pub struct Expr<'a> {
    pub kind: ExprKind<'a>,
    pub line: usize,
    pub col: usize,
}

impl<'a> Expr<'a> {
    /// 构造带位置的表达式
    pub fn new(kind: ExprKind<'a>, line: usize, col: usize) -> Self {
        Expr { kind, line, col }
    }
}

/// 二元运算符枚举
#[derive(Debug, Clone, PartialEq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    EqEq,
    NotEq,
    Lt,
    Le,
    Gt,
    Ge,
    And,
    Or,
    As,
}

File: src/main/kotlin/ast/Expr.kt
package org.pawscript.ast

sealed class Expr : Node {
    data class LiteralInt(val value: Long) : Expr()
    data class LiteralFloat(val value: Double) : Expr()
    data class LiteralBool(val value: Boolean) : Expr()
    data class LiteralChar(val value: Char) : Expr()
    data class LiteralString(val value: String) : Expr()
    data class ArrayLiteral(val elements: List<Expr>) : Expr()
    object Nopaw : Expr()
    data class Variable(val name: String) : Expr()
    data class Binary(val left: Expr, val op: BinaryOp, val right: Expr) : Expr()
    data class Unary(val op: UnaryOp, val expr: Expr) : Expr()
    data class StringTemplate(val parts: List<TemplatePart>) : Expr()
    data class Call(
        val callee: String,
        val typeArgs: List<Type>,
        val positional: List<Expr>,
        val named: Map<String, Expr>
    ) : Expr()

    data class MethodCall(
        val target: Expr,
        val method: String,
        val positional: List<Expr>,
        val named: Map<String, Expr>
    ) : Expr()
    data class FieldAccess(val target: Expr, val field: String) : Expr()
    data class ModuleAccess(val moduleAlias: String, val member: String) : Expr()
    data class As(val expr: Expr, val targetType: Type) : Expr()
    data class Await(val expr: Expr) : Expr()
    data class AskExpr(val prompt: String) : Expr()
    data class Match(
        val scrutinee: Expr,        // 要匹配的目标
        val cases: List<Case>        // 一组 case 分支
    ) : Expr()
    data class Case(
        val pattern: Pattern,
        val result: Expr
    ) : Node
    data class BlockExpr(val statements: List<Statement>) : Expr()
}